Improvements to avx512 adler32 implementations

KungFuJesus opened the pull request:
Now that better benchmarks are in place, it became apparent that masked
broadcast was _not_ faster and it's actually faster to use vmovd, as
suspected.  Additionally, for the VNNI variant, we've unlocked some
additional ILP by doing a second dot product in the loop to a different
running sum that gets recombined later.  This broke a data dependency
chain and allowed the IPC be ~2.75. The result is about a 40-50%
improvement in runtime.

Additionally, we've called the lesser SIMD sized variants if the input
is too small and they happen to be compiled in.  This helps for the
impossibly small input that still is large enough to be a vector length.
For size 16 and 32 inputs I was seeing something like sub 10 ns instead
of 50 ns.

KungFuJesus committed:
Improvements to avx512 adler32 implementations

Now that better benchmarks are in place, it became apparent that masked
broadcast was _not_ faster and it's actually faster to use vmovd, as
suspected.  Additionally, for the VNNI variant, we've unlocked some
additional ILP by doing a second dot product in the loop to a different
running sum that gets recombined later.  This broke a data dependency
chain and allowed the IPC be ~2.75. The result is about a 40-50%
improvement in runtime.

Additionally, we've called the lesser SIMD sized variants if the input
is too small and they happen to be compiled in.  This helps for the
impossibly small input that still is large enough to be a vector length.
For size 16 and 32 inputs I was seeing something like sub 10 ns instead
of 50 ns.

KungFuJesus commented:
```
RUNNING: ./benchmark_zlib_before --benchmark_filter=adler32/avx512 --benchmark_out=/tmp/tmpzb1rrqq0
2022-01-17T09:39:01-05:00
Running ./benchmark_zlib_before
Run on (28 X 4100.33 MHz CPU s)
CPU Caches:
  L1 Data 32 KiB (x14)
  L1 Instruction 32 KiB (x14)
  L2 Unified 1024 KiB (x14)
  L3 Unified 19712 KiB (x1)
Load Average: 0.00, 0.04, 0.06
----------------------------------------------------------------------
Benchmark                            Time             CPU   Iterations
----------------------------------------------------------------------
adler32/avx512/2048               53.5 ns         53.5 ns     13081738
adler32/avx512/4096               92.2 ns         92.2 ns      7597815
adler32/avx512/32768               733 ns          733 ns       954774
adler32/avx512/262144             6366 ns         6366 ns       108722
adler32/avx512/2097152           96446 ns        96444 ns         7240
adler32/avx512/4194304          193522 ns       193517 ns         3690
adler32/avx512_vnni/2048          56.2 ns         56.2 ns     12440116
adler32/avx512_vnni/4096           103 ns          103 ns      6781397
adler32/avx512_vnni/32768          807 ns          807 ns       866709
adler32/avx512_vnni/262144        6479 ns         6478 ns       108032
adler32/avx512_vnni/2097152      78001 ns        77999 ns         8977
adler32/avx512_vnni/4194304     155871 ns       155867 ns         4492
RUNNING: ./benchmark_zlib --benchmark_filter=adler32/avx512 --benchmark_out=/tmp/tmpipyk3a7z
2022-01-17T09:39:12-05:00
Running ./benchmark_zlib
Run on (28 X 3379.89 MHz CPU s)
CPU Caches:
  L1 Data 32 KiB (x14)
  L1 Instruction 32 KiB (x14)
  L2 Unified 1024 KiB (x14)
  L3 Unified 19712 KiB (x1)
Load Average: 0.16, 0.07, 0.07
----------------------------------------------------------------------
Benchmark                            Time             CPU   Iterations
----------------------------------------------------------------------
adler32/avx512/2048               52.4 ns         52.4 ns     13415791
adler32/avx512/4096               91.9 ns         91.9 ns      7634320
adler32/avx512/32768               723 ns          723 ns       967291
adler32/avx512/262144             6476 ns         6475 ns       108028
adler32/avx512/2097152           95760 ns        95758 ns         7274
adler32/avx512/4194304          194399 ns       194395 ns         3600
adler32/avx512_vnni/2048          34.2 ns         34.2 ns     20477196
adler32/avx512_vnni/4096          57.5 ns         57.5 ns     12170323
adler32/avx512_vnni/32768          433 ns          433 ns      1615909
adler32/avx512_vnni/262144        3523 ns         3523 ns       198669
adler32/avx512_vnni/2097152      77722 ns        77720 ns         9004
adler32/avx512_vnni/4194304     155357 ns       155353 ns         4506
Comparing adler32/avx512 (from ./benchmark_zlib_before) to adler32/avx512 (from ./benchmark_zlib)
Benchmark                                                          Time             CPU      Time Old      Time New       CPU Old       CPU New
-----------------------------------------------------------------------------------------------------------------------------------------------
[adler32/avx512 vs. adler32/avx512]/2048                        -0.0208         -0.0208            53            52            53            52
[adler32/avx512 vs. adler32/avx512]/4096                        -0.0028         -0.0028            92            92            92            92
[adler32/avx512 vs. adler32/avx512]/32768                       -0.0131         -0.0131           733           723           733           723
[adler32/avx512 vs. adler32/avx512]/262144                      +0.0172         +0.0172          6366          6476          6366          6475
[adler32/avx512 vs. adler32/avx512]/2097152                     -0.0071         -0.0071         96446         95760         96444         95758
[adler32/avx512 vs. adler32/avx512]/4194304                     +0.0045         +0.0045        193522        194399        193517        194395
[adler32/avx512 vs. adler32/avx512]_vnni/2048                   -0.3922         -0.3922            56            34            56            34
[adler32/avx512 vs. adler32/avx512]_vnni/4096                   -0.4430         -0.4430           103            57           103            57
[adler32/avx512 vs. adler32/avx512]_vnni/32768                  -0.4635         -0.4635           807           433           807           433
[adler32/avx512 vs. adler32/avx512]_vnni/262144                 -0.4562         -0.4562          6479          3523          6478          3523
[adler32/avx512 vs. adler32/avx512]_vnni/2097152                -0.0036         -0.0036         78001         77722         77999         77720
[adler32/avx512 vs. adler32/avx512]_vnni/4194304                -0.0033         -0.0033        155871        155357        155867        155353
OVERALL_GEOMEAN                                                 -0.1783         -0.1783             0             0             0             0
```

_Most_ of the improvements are in VNNI until you get to smaller sized inputs, so really that geometric mean is being poorly represented.

@nmoinvaz Can you check out this PR on your Tigerlake laptop?  I was seeing a 33% improvement in performance with the benchmarks with the one I had tested prior to even making these changes (in comparison to my very expensive, albeit older, HEDT chip).  I'm pretty confident we can probably put the "avx512_well_suited" flag to bed, if so.

mtl1979 commented:
I did kind of say that masked broadcast doesn't seem fastest option... But that was just a hunch and not based on any compiled table...

KungFuJesus commented:
> I did kind of say that masked broadcast doesn't seem fastest option... But that was just a hunch and not based on any compiled table...

Oh for sure, I was only matching the previous behavior that had been there for the loads.  It occurred to me much later that the position the initial separated sums started in made no difference to the final sum.  

What had been confusing to me is once I had made this realization, and I went back to apply it to the AVX512 versions, the previous benchmarks looked slower.  That's because they were actually growing the input size faster and getting stuck on the last input in the timing loop.  And, since the control flow isn't forfeited in that loop until the last invocation of that function is completed, it biases the loop toward a larger input's worth of duration.

Now that we fix the input size, the benchmarks are _much_ more meaningful.  The bigger gains here, though, are due to the improved ILP.

KungFuJesus commented:
Also interesting to note that the gains drop off around the 2-4MB sizes - I'm guessing things stop fitting into cache lines so well there.  Though, it's a purely sequential access pattern, so perhaps it's just the number of times the rebase needs to occur that's killing the gains there.  

There's really no getting around that bit short of reproducing the sequence the compiler is doing on the ALU, and the trip to 64 bit precision and back is much more expensive there (in addition to a half-width 64 bit multiplication).  It's possible there might be some middle ground where we could safely extend NMAX to be a bit larger since we're spread over 16 vector positions but then we're back to trying to rebase every single position in the final sums _before_ a horizontal sum instead of after.  

Mark Adler's carefully selected prime number base strikes again :-/.

mtl1979 commented:
@KungFuJesus If vector instruction sets would have all the features of general purpose registers, we could just accumulate the 32-bit values until the elements contents are slightly less than would overflow from maximum sum of the 8-bit elements... then we would just subtract largest smaller multiple of the base to get everything fit. That way we would only need the trip back to GPR at the end of the loop.

KungFuJesus commented:
> @KungFuJesus If vector instruction sets would have all the features of general purpose registers, we could just accumulate the 32-bit values until the elements contents are slightly less than would overflow from maximum sum of the 8-bit elements... then we would just subtract largest smaller multiple of the base to get everything fit. That way we would only need the trip back to GPR at the end of the loop.

Right, that's what I was suggesting - rebasing the vector register to BASE.  But, doing it requires going to 32 bit->64 bit -> 32 bit.  That alone would be expensive, but the 64 bit low multiplication _really_ kills the gains in doing so.

KungFuJesus commented:
Actually, perf stat suggests it really is the additional cache misses that's killing the improvement.  So perhaps it's just at that point we become LLC bound:
```
adam@eggsbenedict ~/scratch/zlib-ng/build/test/benchmarks $ perf stat -d ./benchmark_zlib --benchmark_filter=adler32/avx512_vnni/2097152
2022-01-17T10:25:47-05:00
Running ./benchmark_zlib
Run on (28 X 3654.81 MHz CPU s)
CPU Caches:
  L1 Data 32 KiB (x14)
  L1 Instruction 32 KiB (x14)
  L2 Unified 1024 KiB (x14)
  L3 Unified 19712 KiB (x1)
Load Average: 0.18, 0.09, 0.08
----------------------------------------------------------------------
Benchmark                            Time             CPU   Iterations
----------------------------------------------------------------------
adler32/avx512_vnni/2097152      81340 ns        81336 ns         8605

 Performance counter stats for './benchmark_zlib --benchmark_filter=adler32/avx512_vnni/2097152':

            864.90 msec task-clock                #    0.973 CPUs utilized          
                 8      context-switches          #    9.250 /sec                   
                 1      cpu-migrations            #    1.156 /sec                   
             2,726      page-faults               #    3.152 K/sec                  
     3,132,704,327      cycles                    #    3.622 GHz                      (62.46%)
     3,256,023,332      instructions              #    1.04  insn per cycle           (75.03%)
       264,953,403      branches                  #  306.339 M/sec                    (75.03%)
           261,892      branch-misses             #    0.10% of all branches          (75.03%)
       423,233,720      L1-dcache-loads           #  489.342 M/sec                    (75.03%)
       318,636,788      L1-dcache-load-misses     #   75.29% of all L1-dcache accesses  (75.03%)
       255,738,654      LLC-loads                 #  295.685 M/sec                    (49.94%)
            12,782      LLC-load-misses           #    0.00% of all LL-cache accesses  (49.94%)

       0.888617020 seconds time elapsed

       0.859313000 seconds user
       0.006002000 seconds sys


adam@eggsbenedict ~/scratch/zlib-ng/build/test/benchmarks $ perf stat -d ./benchmark_zlib --benchmark_filter=adler32/avx512_vnni/32768
2022-01-17T10:26:09-05:00
Running ./benchmark_zlib
Run on (28 X 2142.41 MHz CPU s)
CPU Caches:
  L1 Data 32 KiB (x14)
  L1 Instruction 32 KiB (x14)
  L2 Unified 1024 KiB (x14)
  L3 Unified 19712 KiB (x1)
Load Average: 0.13, 0.08, 0.08
--------------------------------------------------------------------
Benchmark                          Time             CPU   Iterations
--------------------------------------------------------------------
adler32/avx512_vnni/32768        433 ns          433 ns      1617819

 Performance counter stats for './benchmark_zlib --benchmark_filter=adler32/avx512_vnni/32768':

          1,299.19 msec task-clock                #    0.982 CPUs utilized          
                10      context-switches          #    7.697 /sec                   
                 0      cpu-migrations            #    0.000 /sec                   
             3,750      page-faults               #    2.886 K/sec                  
     4,713,668,640      cycles                    #    3.628 GHz                      (62.40%)
    13,439,677,468      instructions              #    2.85  insn per cycle           (74.95%)
       918,425,667      branches                  #  706.922 M/sec                    (75.02%)
         3,087,850      branch-misses             #    0.34% of all branches          (75.06%)
     1,575,187,888      L1-dcache-loads           #    1.212 G/sec                    (75.06%)
       189,367,064      L1-dcache-load-misses     #   12.02% of all L1-dcache accesses  (75.06%)
            64,319      LLC-loads                 #   49.507 K/sec                    (49.92%)
             4,024      LLC-load-misses           #    6.26% of all LL-cache accesses  (49.87%)

       1.322709477 seconds time elapsed

       1.291568000 seconds user
       0.008003000 seconds sys
```

mtl1979 commented:
> Right, that's what I was suggesting - rebasing the vector register to BASE. But, doing it requires going to 32 bit->64 bit -> 32 bit. That alone would be expensive, but the 64 bit low multiplication _really_ kills the gains in doing so.

Multiplication and division is pretty much the most important thing in vector instruction sets that need to improve... Those are way too expensive and as result also modulo is very expensive (essentially division + multiply + subtraction). There should be multiply instructions that don't widen elements. In some cases we know the value can't overflow.

KungFuJesus commented:
Hah, actually...now that I think about it, the smaller input sizes probably need _way_ more passes over the same data to get consistent timing so it probably becomes a self-fulfilling prophecy by that point that nearly everything gets warm in the L1 cache (32K in fact is my and basically everybody else's L1d).  The larger sizes probably would and do benefit from fitting into the L2 and L3's but either don't get warm in there fast enough before the results are consistent from memory.  Though, the minimum number of iterations they run _probably_ has some mitigating effect with that.  

In any case, software prefetching is not going to help with this access pattern (and it basically never helps in my experience, due to the hardware prefetcher basically wanting to do what it wants).  Hitting the memory wall is probably the best we can hope for.  Obviously this could be multithreaded but this library probably doesn't want to do that and it probably makes sense not to so that we can get out of the way for more task level parallelism.

KungFuJesus commented:
> Multiplication and division is pretty much the most important thing in vector instruction sets that need to improve... Those are way too expensive and as result also modulo is very expensive (essentially division + multiply + subtraction). There should be multiply instructions that don't widen elements. In some cases we know the value can't overflow.

Sure, but the fast modulo with a constant trick _requires_ the overflow.  You shift back in the carry bits, so it's basically exactly what we need in this circumstance.  Some of the 32 bit multiplications that occur with the baked in constant could probably be done with pmadd, but yeah it's those overflow bits we want to shift back into the lower half that are the problem.

mtl1979 commented:
@KungFuJesus Under the hood, multiplication is just repeated cycle of shift and addition... Smaller the number is, easier it is to implement with just shift or addition.

Division is on the other saturated subtraction. Basically divisor must be shifted as much left that the most significant 1-bits either align or divisor is one bit to the right and would become larger if shifted to same position.

KungFuJesus commented:
> @KungFuJesus Under the hood, multiplication is just repeated cycle of shift and addition... Smaller the number is, easier it is to implement with just shift or addition.
> 
> Division is on the other saturated subtraction. Basically divisor must be shifted as much left that the most significant 1-bits either align or divisor is one bit to the right and would become larger if shifted to same position.

Right but that's the brute force way.  There's a faster way the compiler is already taking that only requires 1 64 bit multiplication, 1 shift, 1 32 bit multiplication, and a subtraction.  You do this by carefully selecting a divisor based on your constant.  These are tricks libdivide takes for both remainder and division and the compiler does as well, if your modulo or divisor is a constant:

https://lemire.me/blog/2019/02/08/faster-remainders-when-the-divisor-is-a-constant-beating-compilers-and-libdivide/

It's basically a direct solution.

mtl1979 commented:
@KungFuJesus I have read that... It's a neat trick, but still involves division, which my logic didn't... Basically one only needs to know number of leading zero bits of both dividend and divisor... For constants, number of leading zero bits is already know, so don't need to be calculated...

KungFuJesus commented:
> @KungFuJesus I have read that... It's a neat trick, but still involves division, which my logic didn't... Basically one only needs to know number of leading zero bits of both dividend and divisor... For constants, number of leading zero bits is already know, so don't need to be calculated...

Exactly - which we have in this case (BASE).  That's why the compiler is doing this:

```
       mov    $0x80078071,%edi
        ...
        vmovd  %xmm1,%esi 
        mov    %rsi,%rax  // doable in vector registers, not that cheap, spans over 2xvector registers now
        imul   %rdi,%rsi // 64 bit multiplication, pretty expensive in vector registers
        shr    $0x2f,%rsi // cheap in vector registers
        imul   $0xfff1,%esi,%esi // 32 bit multiplication, kind of expensive, but, recombining 2x64 bit registers
                                               // by truncating to 32 bit - very very expensive in vector registers. AVX512 can accomplish
                                               // this with a narrowing vpmovqd in 3 cycles, but the multiplication is still costly
        sub    %esi,%eax // obviously very inexpensive as well
```

In any case, the horizontal addition time isn't killing us here - not really, anyway.  We manage to basically avoid making a trip to stack. It's spilling outside of the cache capacity on the loads that's killing us.

mtl1979 commented:
Not really fan of those magic constants, but always fan of improvements in existing code.... I think I've learnt more about x86 vector instructions in last month than before that since I started studying differences in vector instruction sets (mainly x86, ARM, PPC).

KungFuJesus commented:
Hmm, having some trouble spotting the failure output on the clang debug version.  What does that CI test do that's failing?  Is it in a sanitizer or is somewhere else?

KungFuJesus commented:
Something goofy is happening for me with clang in general.  When trying to build with clang 13 in debug it's failing to link slidehash_avx2.c.o, and instead of object code, for some reason it's still LLVM-IR.

Ahh my bad, that was a result of me messing with LTO options and _not_ turning it off.

KungFuJesus commented:
So I'm still not able to reproduce the clang debug failure, but I only have 12 and 13 to test it on.  The ci is using 11?

Edit: Ok, so I was able to reproduce this with clang, but only on debug?  It seems to be happening in the AVX2 routine of all places, but _only_ when compiled with debug.  This is very strange.

```adam@eggsbenedict ~/scratch/zlib-ng/clang-build $ ./adler32_test 
FAIL [279]: adler32 returned 0x01B21E92 expected 0x01B01E92
```

Exactly 2 off?

KungFuJesus commented:
So something seems to be clobbering the second half of vs2 in debug mode.  Normally I'd say this is the debugger disturbing the state of things, I've seen that happen before (vector register corruption definitely has been a bug in the past with gdb), but I'm seeing the error outside of the debugger.  This is really really strange.  Even stranger is the fact that we didn't see it happen on develop with that AVX2 pull request, since that seems to be the exact spot this issue is happening.

nmoinvaz commented on arch/x86/adler32_avx512.c:
```diff
@@ -15,13 +15,37 @@
 
 #ifdef X86_AVX512_ADLER32
 
+#ifdef X86_AVX2_ADLER32 
```
Have you thought about just including "cpu_features.h"?

KungFuJesus replied:
I could, I figured it made since to limit the scope of things pulled in, but it wouldn't be that much extra work on the compiler.

I think I found the issue btw, though it's strange.  It stems from us extending the xmm register content to a 256 bit wide register.  From what I've read that's supposed to zero extend, however from the looks of it the upper half may be undefined.  I can fix this with a mask of sorts, but I'll see what I can do to make sure this does what it's supposed to.

nmoinvaz replied:
If you want to limit it in scope you could put the externs in adler32_p.h instead. What do you think?

nmoinvaz replied:
Well adler32_p.h is multi-arch so I don't know..

nmoinvaz commented:
### DEVELOP https://github.com/zlib-ng/zlib-ng/commit/30c89988e2399e8da18adedb4ba666e08f078db8
```
PS C:\Users\nathan\Source\zlib-ng\build> .\benchmark_develop.exe --benchmark_filter=adler32/avx*
2022-01-17T10:26:04-08:00
Running C:\Users\nathan\Source\zlib-ng\build\benchmark_develop.exe
Run on (8 X 2995 MHz CPU s)
CPU Caches:
  L1 Data 48 KiB (x4)
  L1 Instruction 32 KiB (x4)
  L2 Unified 1280 KiB (x4)
  L3 Unified 12288 KiB (x1)
----------------------------------------------------------------------
Benchmark                            Time             CPU   Iterations
----------------------------------------------------------------------
adler32/avx2/2048                 46.0 ns         47.1 ns     14933333
adler32/avx2/4096                 79.7 ns         79.5 ns      7466667
adler32/avx2/32768                 596 ns          600 ns      1120000
adler32/avx2/262144               5531 ns         5625 ns       100000
adler32/avx2/2097152             49165 ns        48652 ns        14452
adler32/avx2/4194304             96939 ns        96257 ns         7467
adler32/avx512/2048               47.1 ns         46.0 ns     14933333
adler32/avx512/4096               86.3 ns         87.9 ns      7466667
adler32/avx512/32768               654 ns          656 ns      1000000
adler32/avx512/262144             5286 ns         5301 ns       112000
adler32/avx512/2097152           43777 ns        43493 ns        15448
adler32/avx512/4194304           88119 ns        90332 ns         6400
adler32/avx512_vnni/2048          50.3 ns         51.6 ns     11200000
adler32/avx512_vnni/4096          92.2 ns         92.4 ns      8960000
adler32/avx512_vnni/32768          686 ns          663 ns       896000
adler32/avx512_vnni/262144        5542 ns         5469 ns       100000
adler32/avx512_vnni/2097152      43414 ns        43945 ns        16000
adler32/avx512_vnni/4194304      89006 ns        89979 ns         7467
```
### PR https://github.com/zlib-ng/zlib-ng/pull/1121/commits/5b7f8f4c1d03b269b05470ccbf06aa0fb27ee57d
```
PS C:\Users\nathan\Source\zlib-ng\build> .\benchmark_pr.exe --benchmark_filter=adler32/avx*
2022-01-17T10:26:33-08:00
Running C:\Users\nathan\Source\zlib-ng\build\benchmark_pr.exe
Run on (8 X 2995 MHz CPU s)
CPU Caches:
  L1 Data 48 KiB (x4)
  L1 Instruction 32 KiB (x4)
  L2 Unified 1280 KiB (x4)
  L3 Unified 12288 KiB (x1)
----------------------------------------------------------------------
Benchmark                            Time             CPU   Iterations
----------------------------------------------------------------------
adler32/avx2/2048                 45.9 ns         45.4 ns     14451613
adler32/avx2/4096                 80.1 ns         77.4 ns      7466667
adler32/avx2/32768                 598 ns          600 ns      1120000
adler32/avx2/262144               5506 ns         5469 ns       100000
adler32/avx2/2097152             50454 ns        50000 ns        10000
adler32/avx2/4194304             97138 ns        95215 ns         6400
adler32/avx512/2048               47.2 ns         47.1 ns     14933333
adler32/avx512/4096               88.1 ns         87.9 ns      7466667
adler32/avx512/32768               653 ns          656 ns      1120000
adler32/avx512/262144             5269 ns         5312 ns       100000
adler32/avx512/2097152           44221 ns        43945 ns        16000
adler32/avx512/4194304           88953 ns        87887 ns         7467
adler32/avx512_vnni/2048          31.6 ns         31.4 ns     22400000
adler32/avx512_vnni/4096          51.1 ns         50.0 ns     10000000
adler32/avx512_vnni/32768          381 ns          393 ns      1947826
adler32/avx512_vnni/262144        3031 ns         2982 ns       235789
adler32/avx512_vnni/2097152      37156 ns        36901 ns        19478
adler32/avx512_vnni/4194304      75338 ns        75335 ns        11200
```
So I do see some improvements in `adler32/avx512_vnni` along with more iterations being run.

nmoinvaz commented on arch/x86/adler32_avx512.c:
```diff
@@ -15,13 +15,37 @@
+    /* For impossibly tiny sizes, use the smaller width versions. We still need
+     * to check for compile time support for these but they are likely there */
+#ifdef X86_SSE41_ADLER32 
+    if (len < 32) {
```
I'm not sure if these braces are needed.

KungFuJesus replied:
They aren't

nmoinvaz replied:
I recommend for these changes removing the braces if it is just a one-liner. I say that because other areas and previous code even in adler don't do it.

KungFuJesus commented:
> So I do see some improvements in `adler32/avx512_vnni` along with more iterations being run.

Probably attributed to a larger l2 and l3 I imagine.
Err, only L2.  That, and maybe the hardware prefetcher is smarter in Tigerlake.

KungFuJesus commented:
I figured out the issue, I'm using the wrong intrinsic.  There is a way to convey to the compiler a register to register move so that things get zero extended.

nmoinvaz commented on arch/x86/adler32_avx512.c:
```diff
@@ -15,13 +15,32 @@
+#endif
+
+#ifdef X86_AVX2_ADLER32
+    if (len < 64) {
```
Looks like there was two places with the extra braces. You got the one I didn't mention below.

mtl1979 commented on arch/x86/adler32_avx512_vnni.c:
```diff
@@ -97,38 +131,25 @@ Z_INTERNAL uint32_t adler32_avx512_vnni(uint32_t adler, const unsigned char *buf
     }
-
-    /* Process tail (len < 16).  */
+    
```
Extra whitespace ;)

mtl1979 commented on arch/x86/adler32_avx512.c:
```diff
@@ -73,9 +90,9 @@ Z_INTERNAL uint32_t adler32_avx512(uint32_t adler, const unsigned char *buf, siz
+        vs2 = _mm512_zextsi128_si512(_mm_cvtsi32_si128(sum2));
     }
 
     /* Process tail (len < 16).  */
```
This comment doesn't match VNNI version?

KungFuJesus commented:
@nmoinvaz so it's somewhat interesting you don't see the same improvements in the AVX512 version vs the AVX2 that I do on a Tigerlake laptop with GCC as the compiler on Ubuntu 20.04.  It makes me wonder if MSVC is doing something strange or different there.  Even pre VNNI I was seeing some decent improvements - it was even 30% better than my Cascade Lake 10940X.  I may try this same benchmark on there again to see how much better it gets.

Are you plugged in and using the "performance" mode in windows?

KungFuJesus commented:
```
2022-01-17 14:07:18
Running ./benchmark_zlib
Run on (16 X 4600 MHz CPU s)
CPU Caches:
  L1 Data 48K (x8)
  L1 Instruction 32K (x8)
  L2 Unified 1280K (x8)
  L3 Unified 24576K (x1)
Load Average: 1.45, 0.76, 0.67
----------------------------------------------------------------------
Benchmark                            Time             CPU   Iterations
----------------------------------------------------------------------
adler32/avx2/2048                 53.5 ns         53.5 ns     13517026
adler32/avx2/4096                  103 ns          103 ns      6805283
adler32/avx2/32768                 802 ns          802 ns       872726
adler32/avx2/262144               6740 ns         6740 ns       104439
adler32/avx2/2097152             56505 ns        56504 ns        12449
adler32/avx2/4194304            113094 ns       113095 ns         6207
adler32/avx512/2048               43.4 ns         43.4 ns     16104783
adler32/avx512/4096               81.6 ns         81.6 ns      8542958
adler32/avx512/32768               634 ns          634 ns      1100214
adler32/avx512/262144             5205 ns         5205 ns       136122
adler32/avx512/2097152           44677 ns        44678 ns        15681
adler32/avx512/4194304           89784 ns        89785 ns         7801
adler32/avx512_vnni/2048          28.1 ns         28.1 ns     25124976
adler32/avx512_vnni/4096          47.9 ns         47.9 ns     14628572
adler32/avx512_vnni/32768          360 ns          360 ns      1940455
adler32/avx512_vnni/262144        2973 ns         2973 ns       235838
adler32/avx512_vnni/2097152      37069 ns        37069 ns        18903
adler32/avx512_vnni/4194304      76667 ns        76668 ns         9108
```
MSVC just a suckier compiler?

nmoinvaz commented:
@KungFuJesus yes I am plugged in and with performance mode on. It doesn't make much of a difference between AVX512 and AVX2.

mtl1979 commented:
> MSVC just a suckier compiler?

It is... That's why we suggested Microsoft add zlib-ng to their internal test suite... It's complex enough code to stress test different parts of the compiler...

It might be worth dumping the generated assembly code with gcc, clang and MSVC and send the differing parts to Microsoft so they can improve the optimizer.

nmoinvaz commented on arch/x86/adler32_avx512_vnni.c:
```diff
@@ -11,11 +11,14 @@
 
 #ifdef X86_AVX512VNNI_ADLER32
 
+#include <cpu_features.h>
```
I think this should be "../../cpu_features.h"

nmoinvaz commented on arch/x86/adler32_avx512.c:
```diff
@@ -10,18 +10,36 @@
 
 #ifdef X86_AVX512_ADLER32
 
+#include <cpu_features.h>
```
This should be #include "../../cpu_features.h"

nmoinvaz commented:
> That's why we suggested Microsoft add zlib-ng to their internal test suite...

When was this?

mtl1979 commented:
@nmoinvaz Last time there was issue with MSVC and the intrinsics...

mtl1979 commented on arch/x86/missing_zext.h:
```diff
@@ -0,0 +1,22 @@
+#ifndef MISSING_ZEXT_H
+#define MISSING_ZEXT_H
+
+/* Unfortunately GCC _and_ clang didn't support these things until version
+ * 10 and 12, respectively */
+#include <immintrin.h>
+
+#if (defined(__GNUC__) && __GNUC__ < 10) || (defined(__clang__) && __clang__ < 12)
+static inline __m256i _mm256_zextsi128_si256(__m128i a) {
+    __m128i r;
+    volatile __asm__("vmovdqa %1,%0" : "=x" (r) : "x" (_mm256_castsi256_si128(a)));
+    return _mm256_castsi128_si256(r);
+}
+
+static inline __m512i _mm512_zextsi128_si512(__m128i a) {
+    __m128i r;
+    volatile __asm__("vmovdqa %1,%0" : "=x" (r) : "x" (_mm256_castsi512_si128(a)));
+    return _mm512_castsi128_si512(r);
+}
+#endif
+
+#endif
```
Maybe include these in `/fallback_builtins.h`, so we have one file for all the workarounds for missing intrinsics.

KungFuJesus replied:
Sure.  These are very specific to x86 and GCC and clang.  Hopefully that's not a problem.

mtl1979 commented on fallback_builtins.h:
```diff
@@ -9,6 +9,30 @@
+ * 10 and 12, respectively */
+#include <immintrin.h>
+
+#if (defined(__GNUC__) && __GNUC__ < 10) || (defined(__clang__) && __clang__ < 12)
```
clang also defines `__GNUC__`, so the first test would always be true for clang, as it defines `__GNUC__` as `4`

KungFuJesus replied:
Ugh, of course they do :-/.  Alright, I'll flip those around.

KungFuJesus commented:
> You need to add `!defined(__clang__) &&` to the second part of test as it will still be evaluated for clang...

Man, that file was a bit of a mess.  I renamed the include guard since it didn't actually match the filename.  I also wedged things in a precedence that I _think_ will be correct.

So many #endifs, hah.

KungFuJesus commented:
Ok, pretty sure I fixed most of the issues.  It should compile _and_ work properly.  I tested it by forcing it to be defined by another name in clang 12 and compiling in debug mode.  It does the zero extension move (e.g. vmovdqa xmm3, xmm3), and it seems to have the desired effect (zero'ing out the upper bits).

Interestingly, the same exact assembly is being emitted by GCC when doing the proper intrinsic - I'd have suspected that it could have instead operated on a vector register that it knew had the higher half zeroed, but meh, it's working at least.  No measurable penalty.

nmoinvaz commented on arch/x86/adler32_avx512.c:
```diff
@@ -10,6 +10,8 @@
 
 #include "../../adler32_p.h"
+#include "../../cpu_features.h"
+#include <fallback_builtins.h>
```
We use relative paths for non-system includes.

nmoinvaz commented on arch/x86/adler32_avx2.c:
```diff
@@ -9,6 +9,7 @@
 #include "../../zutil.h"
 
 #include "../../adler32_p.h"
+#include <fallback_builtins.h>
```
Should be relative path..

KungFuJesus commented:
I'm tempted to say the AVX512 being basically the same as the AVX2 is a microsoft issue and remove the "well_suited" flag altogether.  But to support microsoft users, maybe I should just remove the check in front of the vnni function assignment?  We should get more users of all operating systems (and compilers) to run this benchmark with AVX512 capable CPUs to see where we stand on it.  

@nmoinvaz you think you'll be able to try this out on clang or mingw?  I think mingw didn't support AVX512 until waaay later, but it'd be good to know where we sit with this.

KungFuJesus commented:
Oh wow, clang 13 does not optimize that so well, either.  Maybe there's some things I can do to help it along.

KungFuJesus commented:
So I just pushed a commit that helped clang optimize things a bit better, but it's still not near where GCC happens to be.  The good news is that my improvements also showed up as improvements to the GCC implementation, so that's nice:

```
RUNNING: ./benchmark_zlib_before --benchmark_filter=adler32/avx512 --benchmark_out=/tmp/tmpyutndl9n
2022-01-17T18:02:06-05:00
Running ./benchmark_zlib_before
Run on (28 X 3132.39 MHz CPU s)
CPU Caches:
  L1 Data 32 KiB (x14)
  L1 Instruction 32 KiB (x14)
  L2 Unified 1024 KiB (x14)
  L3 Unified 19712 KiB (x1)
Load Average: 0.85, 0.52, 0.36
----------------------------------------------------------------------
Benchmark                            Time             CPU   Iterations
----------------------------------------------------------------------
adler32/avx512/2048               53.5 ns         53.5 ns     13046815
adler32/avx512/4096               92.3 ns         92.3 ns      7579699
adler32/avx512/32768               732 ns          732 ns       952641
adler32/avx512/262144             6441 ns         6441 ns       108644
adler32/avx512/2097152           96437 ns        96433 ns         7286
adler32/avx512/4194304          193063 ns       193055 ns         3623
adler32/avx512_vnni/2048          56.2 ns         56.2 ns     12448045
adler32/avx512_vnni/4096           103 ns          103 ns      6778920
adler32/avx512_vnni/32768          808 ns          808 ns       866399
adler32/avx512_vnni/262144        6479 ns         6479 ns       108024
adler32/avx512_vnni/2097152      77919 ns        77917 ns         8984
adler32/avx512_vnni/4194304     155999 ns       155996 ns         4490
RUNNING: ./benchmark_zlib --benchmark_filter=adler32/avx512 --benchmark_out=/tmp/tmpobjgi3u4
2022-01-17T18:02:17-05:00
Running ./benchmark_zlib
Run on (28 X 4100.03 MHz CPU s)
CPU Caches:
  L1 Data 32 KiB (x14)
  L1 Instruction 32 KiB (x14)
  L2 Unified 1024 KiB (x14)
  L3 Unified 19712 KiB (x1)
Load Average: 1.04, 0.57, 0.38
----------------------------------------------------------------------
Benchmark                            Time             CPU   Iterations
----------------------------------------------------------------------
adler32/avx512/2048               42.4 ns         42.4 ns     16430902
adler32/avx512/4096               73.0 ns         73.0 ns      9636179
adler32/avx512/32768               575 ns          575 ns      1223310
adler32/avx512/262144             4994 ns         4994 ns       140257
adler32/avx512/2097152           86403 ns        86400 ns         8094
adler32/avx512/4194304          172886 ns       172882 ns         4049
adler32/avx512_vnni/2048          34.3 ns         34.3 ns     20421006
adler32/avx512_vnni/4096          57.6 ns         57.6 ns     12145163
adler32/avx512_vnni/32768          434 ns          434 ns      1613121
adler32/avx512_vnni/262144        3495 ns         3495 ns       200251
adler32/avx512_vnni/2097152      77588 ns        77586 ns         9019
adler32/avx512_vnni/4194304     155451 ns       155448 ns         4504
Comparing adler32/avx512 (from ./benchmark_zlib_before) to adler32/avx512 (from ./benchmark_zlib)
Benchmark                                                          Time             CPU      Time Old      Time New       CPU Old       CPU New
-----------------------------------------------------------------------------------------------------------------------------------------------
[adler32/avx512 vs. adler32/avx512]/2048                        -0.2076         -0.2076            53            42            53            42
[adler32/avx512 vs. adler32/avx512]/4096                        -0.2089         -0.2089            92            73            92            73
[adler32/avx512 vs. adler32/avx512]/32768                       -0.2148         -0.2148           732           575           732           575
[adler32/avx512 vs. adler32/avx512]/262144                      -0.2247         -0.2247          6441          4994          6441          4994
[adler32/avx512 vs. adler32/avx512]/2097152                     -0.1040         -0.1040         96437         86403         96433         86400
[adler32/avx512 vs. adler32/avx512]/4194304                     -0.1045         -0.1045        193063        172886        193055        172882
[adler32/avx512 vs. adler32/avx512]_vnni/2048                   -0.3906         -0.3906            56            34            56            34
[adler32/avx512 vs. adler32/avx512]_vnni/4096                   -0.4415         -0.4415           103            58           103            58
[adler32/avx512 vs. adler32/avx512]_vnni/32768                  -0.4628         -0.4628           808           434           808           434
[adler32/avx512 vs. adler32/avx512]_vnni/262144                 -0.4606         -0.4607          6479          3495          6479          3495
[adler32/avx512 vs. adler32/avx512]_vnni/2097152                -0.0042         -0.0042         77919         77588         77917         77586
[adler32/avx512 vs. adler32/avx512]_vnni/4194304                -0.0035         -0.0035        155999        155451        155996        155448
OVERALL_GEOMEAN                                                 -0.2550         -0.2551             0             0             0             0
```

KungFuJesus commented:
And if we needed any further confirmation we've hit the memory wall, here's isa-l's internal distribution of igzip's adler32 benchmark with us patched in, cold (larger than cache), and hot (fits within cache):

```
./adler32_perf 
adler32_perf:
adler32_cold: runtime =    3041478 usecs, bandwidth 33151 MB in 3.0415 sec = 10899.89 MB/s
adler32-zlib-ng_cold: runtime =    3066625 usecs, bandwidth 46976 MB in 3.0666 sec = 15318.53 MB/s
adler32-ipp_cold: runtime =    3060747 usecs, bandwidth 26508 MB in 3.0607 sec = 8660.63 MB/s

./adler32_perf 
adler32_perf:
adler32_warm: runtime =    3062236 usecs, bandwidth 65046 MB in 3.0622 sec = 21241.55 MB/s
adler32-zlib-ng_warm: runtime =    3062876 usecs, bandwidth 228412 MB in 3.0629 sec = 74574.45 MB/s
adler32-ipp_warm: runtime =    3062549 usecs, bandwidth 37912 MB in 3.0625 sec = 12379.46 MB/s

```

mtl1979 commented on fallback_builtins.h:
```diff
@@ -41,4 +41,27 @@ static __forceinline unsigned long long __builtin_ctzll(uint64_t value) {
+#endif // __AVX2__
+
+#endif // clang/gcc test
+#endif // include guard fallback_builtin_h
```
Should be `FALLBACK_BUILTINS_H` to match top of file. Missing `S` in the end, not capitalized.

KungFuJesus commented:
Hah...well that's rich.  When trying to figure out why clang was doing so much worse for AVX512 vs VNNI, I discovered that -march=native was on and ironically, it removed one of the independent vpadds from the vpmadd operations and replaced it with VNNI dot product instruction.  Which yeah in most circumstances might have been better, but because we didn't have the extra accumulation with a separate sum for one of those dot products with an unrolled by 2 loop like we do in the VNNI _explicit_ version, it was net slower.  Turning off -march=native optimization brought the runtime basically in line with what GCC was producing.  

@nmoinvaz I sort of wonder if MSVC is taking similar liberties.  You can't control what instructions end up being emitted nearly as well, though Microsoft swears up and down it won't use anything but the baseline instructions for AVX512 unless you use their intrinsics.  I dunno, could be part of what's going on.  In any case, definitely try clang on your windows laptop, I suspect you'll see better AVX512 than AVX2 numbers.  And obviously, AVX512-VNNI numbers better than that.

nmoinvaz commented:
## ClangCl Benchmarks  (Laptop NOT plugged in)

### DEVELOP https://github.com/zlib-ng/zlib-ng/commit/30c89988e2399e8da18adedb4ba666e08f078db8
```
PS C:\Users\nathan\Source\zlib-ng\build> .\benchmark_develop_clang.exe --benchmark_filter=adler32/avx*
2022-01-17T16:12:18-08:00
Running C:\Users\nathan\Source\zlib-ng\build\benchmark_develop_clang.exe
Run on (8 X 2995 MHz CPU s)
CPU Caches:
  L1 Data 48 KiB (x4)
  L1 Instruction 32 KiB (x4)
  L2 Unified 1280 KiB (x4)
  L3 Unified 12288 KiB (x1)
----------------------------------------------------------------------
Benchmark                            Time             CPU   Iterations
----------------------------------------------------------------------
adler32/avx2/2048                 82.8 ns         82.0 ns      8960000
adler32/avx2/4096                  141 ns          140 ns      4480000
adler32/avx2/32768                1044 ns         1050 ns       640000
adler32/avx2/262144               9926 ns        10045 ns        74667
adler32/avx2/2097152             89900 ns        90681 ns         8960
adler32/avx2/4194304            156708 ns       156941 ns         4978
adler32/avx512/2048               73.6 ns         72.5 ns     11200000
adler32/avx512/4096                135 ns          137 ns      5600000
adler32/avx512/32768              1031 ns         1001 ns       640000
adler32/avx512/262144             8639 ns         8545 ns        89600
adler32/avx512/2097152           73657 ns        73242 ns         8960
adler32/avx512/4194304          148756 ns       153460 ns         5600
adler32/avx512_vnni/2048          79.3 ns         80.2 ns      8960000
adler32/avx512_vnni/4096           141 ns          138 ns      4977778
adler32/avx512_vnni/32768         1138 ns         1123 ns       640000
adler32/avx512_vnni/262144        9122 ns         9208 ns        74667
adler32/avx512_vnni/2097152      72793 ns        73940 ns        11200
adler32/avx512_vnni/4194304     149194 ns       150663 ns         4978
```
### PR https://github.com/zlib-ng/zlib-ng/pull/1121/commits/ad0cb0cc19b760143eec23a97b45fa54341d578c
```
PS C:\Users\nathan\Source\zlib-ng\build> .\benchmark_pr_clang.exe --benchmark_filter=adler32/avx*
2022-01-17T16:12:49-08:00
Running C:\Users\nathan\Source\zlib-ng\build\benchmark_pr_clang.exe
Run on (8 X 2995 MHz CPU s)
CPU Caches:
  L1 Data 48 KiB (x4)
  L1 Instruction 32 KiB (x4)
  L2 Unified 1280 KiB (x4)
  L3 Unified 12288 KiB (x1)
----------------------------------------------------------------------
Benchmark                            Time             CPU   Iterations
----------------------------------------------------------------------
adler32/avx2/2048                 82.6 ns         82.0 ns      8960000
adler32/avx2/4096                  154 ns          157 ns      4977778
adler32/avx2/32768                1010 ns         1001 ns       640000
adler32/avx2/262144               9405 ns         9417 ns        89600
adler32/avx2/2097152             90908 ns        92425 ns         8960
adler32/avx2/4194304            159180 ns       160435 ns         4480
adler32/avx512/2048               51.3 ns         51.6 ns     10000000
adler32/avx512/4096               90.5 ns         92.1 ns      7466667
adler32/avx512/32768               657 ns          642 ns      1120000
adler32/avx512/262144             5597 ns         5580 ns       112000
adler32/avx512/2097152           59136 ns        58594 ns        11200
adler32/avx512/4194304          119273 ns       119275 ns         4978
adler32/avx512_vnni/2048          49.2 ns         50.0 ns     10000000
adler32/avx512_vnni/4096          80.0 ns         80.2 ns      8960000
adler32/avx512_vnni/32768          580 ns          572 ns      1120000
adler32/avx512_vnni/262144        4841 ns         4956 ns       154483
adler32/avx512_vnni/2097152      59935 ns        59989 ns        11200
adler32/avx512_vnni/4194304     123412 ns       124512 ns         6400
```

So with ClangCl I am seeing some improvements in AVX512 and also some increase in iterations in VNNI max size with same time as AVX512.

I also found an issue with CMake using ClangCl not properly detecting VNNI so I will submit a PR for that.

KungFuJesus commented:
> So with ClangCl I am seeing some improvements in AVX512 and also some increase in iterations in VNNI max size with same time as AVX512.
> 
> I also found an issue with CMake using ClangCl not properly detecting VNNI so I will submit a PR for that.

I wouldn't pay _too_ much attention to the number of iterations.  It loosely correlates to the standard deviation or how many different times the loop needs to be run to get something remotely consistent.  If it's absurdly high it likely gave up trying to get a consistent average.  However, for something that completes very quickly google benchmark also has a minimal amount of time it wants to average samples over, so it will be higher in those scenarios as well.

nmoinvaz commented:
I forgot I am not plugged in. Let me rerun those benchmarks. Dear lord!

nmoinvaz commented:
## ClangCl Benchmarks (Laptop plugged in)
### DEVELOP https://github.com/zlib-ng/zlib-ng/commit/30c89988e2399e8da18adedb4ba666e08f078db8
```
PS C:\Users\nathan\Source\zlib-ng\build> .\benchmark_develop_clang.exe --benchmark_filter=adler32/avx*
2022-01-17T16:36:09-08:00
Running C:\Users\nathan\Source\zlib-ng\build\benchmark_develop_clang.exe
Run on (8 X 2995 MHz CPU s)
CPU Caches:
  L1 Data 48 KiB (x4)
  L1 Instruction 32 KiB (x4)
  L2 Unified 1280 KiB (x4)
  L3 Unified 12288 KiB (x1)
----------------------------------------------------------------------
Benchmark                            Time             CPU   Iterations
----------------------------------------------------------------------
adler32/avx2/2048                 56.2 ns         56.2 ns     10000000
adler32/avx2/4096                 99.6 ns         98.4 ns      7466667
adler32/avx2/32768                 692 ns          680 ns       988690
adler32/avx2/262144               6353 ns         6417 ns       112000
adler32/avx2/2097152             54371 ns        54688 ns        10000
adler32/avx2/4194304            106683 ns       108817 ns         5600
adler32/avx512/2048               47.1 ns         47.4 ns     13841655
adler32/avx512/4096               89.3 ns         87.9 ns      7466667
adler32/avx512/32768               659 ns          663 ns       896000
adler32/avx512/262144             5844 ns         5755 ns        89600
adler32/avx512/2097152           43790 ns        43945 ns        16000
adler32/avx512/4194304           94359 ns        95912 ns         8960
adler32/avx512_vnni/2048          52.7 ns         53.1 ns     10000000
adler32/avx512_vnni/4096          89.3 ns         87.9 ns      6400000
adler32/avx512_vnni/32768          683 ns          698 ns      1120000
adler32/avx512_vnni/262144        5870 ns         5938 ns       100000
adler32/avx512_vnni/2097152      45019 ns        44922 ns        16000
adler32/avx512_vnni/4194304      87445 ns        87193 ns         8960
```
### PR https://github.com/zlib-ng/zlib-ng/commit/ad0cb0cc19b760143eec23a97b45fa54341d578c
```

PS C:\Users\nathan\Source\zlib-ng\build> .\benchmark_pr_clang.exe --benchmark_filter=adler32/avx*
2022-01-17T16:36:33-08:00
Running C:\Users\nathan\Source\zlib-ng\build\benchmark_pr_clang.exe
Run on (8 X 2995 MHz CPU s)
CPU Caches:
  L1 Data 48 KiB (x4)
  L1 Instruction 32 KiB (x4)
  L2 Unified 1280 KiB (x4)
  L3 Unified 12288 KiB (x1)
----------------------------------------------------------------------
Benchmark                            Time             CPU   Iterations
----------------------------------------------------------------------
adler32/avx2/2048                 57.2 ns         55.8 ns      8960000
adler32/avx2/4096                 97.9 ns         98.4 ns      7466667
adler32/avx2/32768                 706 ns          698 ns       896000
adler32/avx2/262144               6320 ns         6452 ns        89600
adler32/avx2/2097152             54416 ns        56250 ns        10000
adler32/avx2/4194304            106717 ns       106027 ns         5600
adler32/avx512/2048               32.6 ns         33.0 ns     20363636
adler32/avx512/4096               59.5 ns         57.2 ns     11200000
adler32/avx512/32768               433 ns          435 ns      1723077
adler32/avx512/262144             3560 ns         3610 ns       194783
adler32/avx512/2097152           40156 ns        39237 ns        17920
adler32/avx512/4194304           77187 ns        78474 ns         8960
adler32/avx512_vnni/2048          32.5 ns         33.0 ns     21333333
adler32/avx512_vnni/4096          56.8 ns         57.8 ns     10000000
adler32/avx512_vnni/32768          386 ns          384 ns      1792000
adler32/avx512_vnni/262144        2986 ns         3048 ns       235789
adler32/avx512_vnni/2097152      37009 ns        36830 ns        18667
adler32/avx512_vnni/4194304      76852 ns        76730 ns         8960
```

KungFuJesus commented:
That is much more in line with what I'd expect.  Think we can kill avx512_well_suited?

nmoinvaz commented:
Yes I think so... AVX512 now appears to aways be faster than AVX2. And AVX512VINNI is at least as far as AVX512 or faster even when not plugged in. I would make those changes as a second commit.

KungFuJesus commented:
Did CI take a vacation or something?  What's up with the queue?

nmoinvaz commented:
I think GitHub probably throttles us sometimes so we don't monopolize the resources.

KungFuJesus commented:
Parent company must have heard us talking badly about their compiler, hah.

mtl1979 commented:
> Parent company must have heard us talking badly about their compiler, hah.

Didn't you see one of their GitHub accounts already posting "elsewhere"... I pretty much said the issue is not in zlib-ng, but their compiler suite. Kinda not saying "Can't fix stupid"... eh... but anyways... We can continue that discussion there ;)

mtl1979 commented:
> I think GitHub probably throttles us sometimes so we don't monopolize the resources.

I'm assuming we're still using their free tier... We do pull requests in quite fast succession when something doesn't work and CI tests fail. They could easily cancel previous commits, but even that takes resources when the runners need to reset.

KungFuJesus commented:
Ok, those are the 3 checks I expect to fail. Looks good now.

Dead2 commented on arch/x86/adler32_avx512.c:
```diff
@@ -33,52 +71,54 @@ Z_INTERNAL uint32_t adler32_avx512(uint32_t adler, const unsigned char *buf, siz
+    const __m512i zero = _mm512_setzero_si512();
 
     while (len >= 64) {
-       __m512i vs1_0 = vs1;
```
Looks like there are a lot of whitespace changes. Github doesn't show whether this is tab or space, but we use 4-space indentation, hope you make sure that is correct.

KungFuJesus replied:
They're all 4 space tabs.  A lot of the original implementations these were originally copied from actually had inconsistent indentation (the ssse3 version seemed to be the original culprit).

mtl1979 replied:
The first underscore should align with `e`, not `l`... Not sure if it's rendering glitch or not...

KungFuJesus replied:
![spacing](https://user-images.githubusercontent.com/3620277/149960561-211ce986-19af-4b3c-a62b-cba213b2d90e.png)

mtl1979 replied:
Sometimes I look at the actual file, not the diff to see everything is OK... I've seen many diffs render whitespace incorrectly especially when TABs and SPACEs are mixed in same text block.

KungFuJesus replied:
That is the actual file.  My vimrc doesn't place tabs and I've only ever edited that file on this system.  
```
set expandtab
set tabstop=4
set shiftwidth=4
set smarttab
set nocp
```
`:retab` makes no difference, they're definitely spaces if they weren't already.  In this case the diff is showing where I fixed what the indentation was, as I think that bit managed to get copied from the previous adler32 implementations.

KungFuJesus replied:
Here's the patient zero commit: https://github.com/zlib-ng/zlib-ng/blame/develop/arch/x86/adler32_ssse3.c#L54

mtl1979 replied:
I guess the SSSE3 version should be cleaned up in separate pull request...

Dead2 commented on arch/x86/x86.c:
```diff
@@ -28,7 +28,6 @@ Z_INTERNAL int x86_cpu_has_sse42;
 Z_INTERNAL int x86_cpu_has_pclmulqdq;
 Z_INTERNAL int x86_cpu_has_vpclmulqdq;
 Z_INTERNAL int x86_cpu_has_tzcnt;
-Z_INTERNAL int x86_cpu_well_suited_avx512;
```
I would prefer that the removal of cpu_well_suited_avx512 stuff is in a separate commit.

KungFuJesus committed:
Remove the "avx512_well_suited" cpu flag

Now that we have confirmation that the AVX512 variants so far have been
universally better on every capable CPU we've tested them on, there's no
sense in trying to maintain a whitelist.

mtl1979 commented on arch/x86/adler32_avx512.c:
```diff
@@ -10,18 +10,56 @@
+     * to check for compile time support for these but they are likely there */
+#ifdef X86_SSE41_ADLER32 
+    if (len < 32) 
+       return adler32_sse41(adler, buf, len);
```
4 space indentation...

mtl1979 commented on arch/x86/adler32_avx512.c:
```diff
@@ -10,18 +10,56 @@
+
+#ifdef X86_AVX2_ADLER32
+    if (len < 64)
+       return adler32_avx2(adler, buf, len);
```
4 space indentation...

nmoinvaz commented on arch/x86/x86.c:
```diff
@@ -61,12 +60,12 @@ static void cpuidex(int info, int subinfo, unsigned* eax, unsigned* ebx, unsigne
 
     cpuid(0, &maxbasic, &ebx, &ecx, &edx);
 
+    /* In case this ever needs to be dug up again, this is
```
Maybe it is best to just remove this, we can always look it up in Git history..

KungFuJesus replied:
I suppose that's true, you just have to know where the body is buried.

KungFuJesus commented:
Just pushed some minor corrections to some comments, nothing functionally different

Dead2 commented on arch/x86/x86.c:
```diff
@@ -61,32 +60,14 @@ static void cpuidex(int info, int subinfo, unsigned* eax, unsigned* ebx, unsigne
-    char cpu_identity[13];
 
     cpuid(0, &maxbasic, &ebx, &ecx, &edx);
 
```
Just wondering about the commit description
```
The attributes that are extracted from the cpuid instruction are useful
and bear keeping around, if not just in a commented out state.  There
may be a time when extracting that model info or AMD vs intel is important.
```
Is that accurate? It looks to me like it gets removed in this commit..

KungFuJesus replied:
Ah yes, I'll amend the commit message. It was suggested in review at some point that their existence in version control is probably enough that they don't need to be commented out code blocks.

Dead2 reviewed (approved):
LGTM
